llvm/lib/CAS/MappedFileRegionBumpPtr.cpp

```diff
@@ -0,0 +1,284 @@
+/// implementation strategy is:
+///
+/// 1. Use \c ftruncate (\c sys::fs::resize_file) to grow the file to its max
+///    size (typically several GB). Many modern filesystems will create a sparse
+///    file, so that the trailing unused pages do not take space on disk.
+/// 2. Call \c mmap (\c sys::fs::mapped_file_region)
+/// 3. [Automatic as part of 2.]
+/// 4. [Automatic as part of 2.]
+///
+/// Additionally, we attempt to resize the file to its actual data size when
+/// closing the mapping, if this is the only concurrent instance. This is done
+/// using file locks. Not implemented on Windows.
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CAS/MappedFileRegionBumpPtr.h"
+#include "llvm/ADT/ScopeExit.h"
+#include "llvm/ADT/StringMap.h"
+#include <mutex>
+
+#if LLVM_ENABLE_ONDISK_CAS
+using namespace llvm;
+using namespace llvm::cas;
+
+namespace {
+struct FileLockRAII {
+  std::string Path;
+  int FD;
+  enum LockKind { Shared, Exclusive };
+  std::optional<LockKind> Locked;
+
+  FileLockRAII(std::string Path, int FD) : Path(std::move(Path)), FD(FD) {}
+  ~FileLockRAII() { consumeError(unlock()); }
+
+  Error lock(LockKind LK) {
+    if (std::error_code EC = sys::fs::lockFile(FD, LK == Exclusive))
+      return createFileError(Path, EC);
+    Locked = LK;
+    return Error::success();
+  }
+
+  /// Returns true if the lock was acquired.
+  bool tryLockExclusive(std::chrono::milliseconds Timeout) {
+    if (sys::fs::tryLockFile(FD, Timeout) == std::error_code()) {
+      Locked = Exclusive;
+      return true;
+    }
+    return false;
+  }
+
+  Error unlock() {
+    if (Locked) {
+      Locked = std::nullopt;
+      if (std::error_code EC = sys::fs::unlockFile(FD))
+        return createFileError(Path, EC);
+    }
+    return Error::success();
+  }
+
+  /// Take responsibility for unlocking the lock.
+  void takeLock() {
+    assert(Locked);
+    Locked = std::nullopt;
+  }
+};
+} // end anonymous namespace
+
+Expected<MappedFileRegionBumpPtr> MappedFileRegionBumpPtr::create(
+    const Twine &Path, uint64_t Capacity, int64_t BumpPtrOffset,
+    function_ref<Error(MappedFileRegionBumpPtr &)> NewFileConstructor) {
+  MappedFileRegionBumpPtr Result;
+  Result.Path = Path.str();
+  int FD;
+  if (std::error_code EC = sys::fs::openFileForReadWrite(
+          Result.Path, FD, sys::fs::CD_OpenAlways, sys::fs::OF_None))
+    return errorCodeToError(EC);
+  Result.FD = FD;
+
+  sys::fs::file_t File = sys::fs::convertFDToNativeFile(FD);
+
+  // Use shared/reader locking in case another process is in the process of
+  // initializing the file.
+  FileLockRAII FileLock(Result.Path, FD);
+  if (Error E = FileLock.lock(FileLockRAII::Shared))
+    return std::move(E);
+
+  sys::fs::file_status Status;
+  if (std::error_code EC = sys::fs::status(File, Status))
+    return createFileError(Result.Path, EC);
+
+  while (Status.getSize() < Capacity) {
```

Comment:
To make sure I understand, you're proposing:

Open
* Take shared lock on (1) -- you are now guaranteed no one will shrink the file, but someone could grow it.
* Check if the size is correct; if so we're done.  Otherwise:
* Take exclusive lock on (2)
* Initialize or already done by racing process
* Unlock (2)

Close (you still have shared lock (1))
* Try to promote to exclusive lock on (1)
* Shrink if you get exclusive access
* Unlock (1)

One thing I realized while thinking this through is that there is an assumption that everyone accessing the same file will have the same `Capacity` value.  In my current implementation if someone initializes the file with capacity X, and another process wants capacity X+1 it will spin in the loop trying to get exclusive access to grow the file until the first process closes the mapping.  In your proposed locking scheme IIUC we would grow the file while another process is already mapping it with a smaller capacity, which is fine as long as we never write data past the smaller capacity, but if we do then it can have out of bound offsets.  Currently we can't detect this, because we can't tell the difference between "file is in use by another process but with smaller capacity than ours" and "file is in shrunken state".  

Not sure this is worth worrying about since we expect the CAS to use fixed capacity outside of testing situations where we use the env vars.  We could also solve it by storing information in the file itself; e.g. we could store the capacity the first time it was opened and always use that after it's created.

Reply:
> we can't tell the difference between "file is in use by another process but with smaller capacity than ours" and "file is in shrunken state".

Couldn't we compare the "used size"/BumpPtr against the file size? In any case, I'd suggest adding a comment noting this but otherwise we don't need to worry until we allow clients to configure the capacity.

Reply:
> Couldn't we compare the "used size"/BumpPtr against the file size?

Technically you could have the edge case that
* file is opened by process 1 and is filled until it's exactly at process 1's capacity
* file is opened by process 2 with a larger capacity, but it has size == bumpptr, so we think no one has it open

But yeah, seems okay for now.